### PR TITLE
Lower bound in scan minimizer

### PR DESCRIPTION
**Pull request by HansN87**
Solves https://github.com/IceCubeOpenSource/skyllh/issues/32

### CODE REVIEW

File: skyllh/core/minimizer.py [hunk 1]
```diff
@@ -366,37 +368,39 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
...
+        f = None
         at_boundary = False
 
         # We do the minimization process while the precision of ns is not
         # reached yet or the function is still rising or falling fast, i.e. the
         # minimum is in a deep well.
-        while ((ns_tol < np.fabs(step)) or (np.fabs(fprime) > 1) and niter<max_steps): 
+        # In case the optimum is found outside the bounds on ns the best fit
+        # will be set to the boundary value.
+        while ((ns_tol < np.fabs(step)) or (np.fabs(fprime) > 1) and niter < max_steps):
```
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: https://github.com/IceCubeOpenSource/skyllh/commit/bc96a042b0ab6673e53b9f497c8ce64ede7bbaf8

File: skyllh/core/minimizer.py [hunk 2]
```diff
@@ -366,37 +368,39 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
         # Initialize stepsize to be larger than ns tolerance.
         # Also initialize first derivative to large value.
         # Want to perform at least one NR iteration.
-        step = ns_tol + 1 
+        step = ns_tol + 1
         fprime = 1000
-        # NR does not guarantee, thus limit iterations.
+        # NR does not guarantee convergence, thus limit iterations.
         max_steps = 100
         status = {'warnflag': 0, 'warnreason': ''}
```
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: https://github.com/IceCubeOpenSource/skyllh/commit/bc96a042b0ab6673e53b9f497c8ce64ede7bbaf8

File: skyllh/core/minimizer.py [hunk 3]
```diff
@@ -366,37 +368,39 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
         # Initialize stepsize to be larger than ns tolerance.
         # Also initialize first derivative to large value.
         # Want to perform at least one NR iteration.
-        step = ns_tol + 1 
+        step = ns_tol + 1
         fprime = 1000
-        # NR does not guarantee, thus limit iterations.
+        # NR does not guarantee convergence, thus limit iterations.
         max_steps = 100
```
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: https://github.com/IceCubeOpenSource/skyllh/commit/bc96a042b0ab6673e53b9f497c8ce64ede7bbaf8

File: skyllh/core/minimizer.py [hunk 4]
```diff
@@ -366,37 +368,39 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
...
             step = -fprime / fprimeprime
 
-            # Treat boundary values.
+            # Exit optimization if ns is at boundary but next step would be outside.
             if((ns == ns_min and step < 0.0) or (ns == ns_max and step > 0.0)):
                 at_boundary = True
                 f_sol = f
 
                 if(ns == ns_min):
                     status['warnreason'] = 'Function minimum is below the '\
```
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: https://github.com/IceCubeOpenSource/skyllh/commit/bc96a042b0ab6673e53b9f497c8ce64ede7bbaf8

File: skyllh/core/minimizer.py [hunk 5]
```diff
@@ -408,20 +412,20 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
...
         x[0] = ns
         # Once converged evaluate function at minimum unless hit boundary
         if(not at_boundary):
-            (f_sol, fprime, fprimeprime) = func(x, *func_args) 
+            (f_sol, fprime, fprimeprime) = func(x, *func_args)
 
-        if(niter==max_steps):
+        if(niter == max_steps):
             status['warnflag'] = 1
             status['warnreason'] = 'NR optimization did not converge within {}'\
```
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: https://github.com/IceCubeOpenSource/skyllh/commit/bc96a042b0ab6673e53b9f497c8ce64ede7bbaf8

File: skyllh/core/minimizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (HansN87) [reply skyllh/core/minimizer.py]: done

File: skyllh/core/minimizer.py [hunk 6]
```diff
@@ -354,58 +361,79 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
...
+        # Once converged evaluate function at minimum value unless
+        # Convergence was forced at boundary
+        # in which case function value is already known
+        if(not at_boundary):
+            (f, fprime, fprimeprime) = func(x, *func_args)
+
+        if(niter == max_steps):
+            status['warnflag'] = 1
+            status['warnreason'] = ('NR optimization did not converge within {} '
+                                   'NR steps.'.format(niter))
```
Comment (martwo) [inline skyllh/core/minimizer.py]:
```suggestion
                                    'NR steps.'.format(niter))
```

File: skyllh/core/minimizer.py [hunk 7]
```diff
@@ -336,14 +342,15 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
                 The warning flag indicating if the minimization did converge.
                 The possible values are:
 
+                    -1: The function minimum is above the upper bound of the
+                       parameter value. Convergence forced at upper bound.
+                    -2: The function minimum is below the lower bound of the
+                       parameter value. Convergence forced at lower bound.
```
Comment (martwo) [inline skyllh/core/minimizer.py]:
```suggestion
                    -1: The function minimum is above the upper bound of the
                        parameter value. Convergence forced at upper bound.
                    -2: The function minimum is below the lower bound of the
                        parameter value. Convergence forced at lower bound.
```

File: skyllh/core/minimizer.py [hunk 8]
```diff
@@ -336,14 +342,15 @@ def minimize(self, initials, bounds, func, func_args=None, **kwargs):
...
                     0: The minimization converged with a iteration step size
                        smaller than the specified precision.
-                    1: The function minimum is below the minimum bound of the
-                       parameter value. The last iteration's step size did not
-                       achieve the specified precision.
-                    2: The function minimum is above the maximum bound of the
-                       parameter value. The last iteration's step size did not
-                       achieve the specified precision.
+                    1: The minimization did NOT converge within self.max_steps
+                        number of steps
```
Comment (martwo) [inline skyllh/core/minimizer.py]:
```suggestion
                       number of steps
```

File: skyllh/core/minimizer.py [hunk 9]
```diff
@@ -273,18 +274,23 @@ class NR1dNsMinimizerImpl(MinimizerImpl):
...
+    def __init__(self, ns_tol=1e-3, max_steps=100):
         """Creates a new NRNs minimizer instance to minimize the given
         likelihood function with its given partial derivatives.
 
         Parameters
         ----------
         ns_tol : float
             The tolerance / precision for the ns parameter value.
+        max_steps : int
+            The maximum number NR steps. If max_step is reached,
```
Comment (martwo) [inline skyllh/core/minimizer.py]:
```suggestion
            The maximum number of NR steps. If max_step is reached,
```

**Pull request merged**